File: 2007. Find Original Array From Doubled Array/Solution2007.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class Solution2007 {
public:
    vector<int> findOriginalArray(vector<int>& changed) {
        if (changed.size() % 2 != 0) return {};
        int cmax = changed[0];
        for (int c : changed) {
            cmax = max(cmax, c);
        }
        vector<int> freq(cmax*2+1, 0);
        for (int c : changed) {
            freq[c]++;
        }
        vector<int> original;
        for (int num = 0; num <= cmax; num++) {
            if (freq[num]) {
                freq[num]--;
                int dnum = 2 * num;
                if (freq[dnum] > 0) {
                    freq[dnum]--;
                    original.push_back(num);
                    num--; //if multiple num exists
                } else {
                    return {};
                }
            }
        }
        return original;
    }
};

File: 2007. Find Original Array From Doubled Array/Solution2007.py
from collections import Counter
from typing import List


class Solution2007:
    def findOriginalArray(self, changed: List[int]) -> List[int]:
        if len(changed) % 2 != 0:
            return []
        original = []
        max_changed = max(changed)
        freq = Counter(changed)
            
        for i in range(max_changed+1):
            if i == 0:
                if freq[i] % 2 > 0:
                    return []
                original += [0] * (freq[i] // 2)
            elif freq[i] > 0: 
                original += [i] * freq[i]
                if freq[2*i] >= freq[i]:
                    freq[2*i]-=freq[i]
                else:
                    return []
        return original